chain/store/name.go

```diff
@@ -1,63 +1,178 @@
 	if err != nil {
 		return err
 	}
+	ni.expiresAt, err = serialization.ReadUint32(r)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-	return sdb.trie.TryUpdate(append(NameRegistrantPrefix, registrantId...), []byte(name))
+func (ni *nameInfo) Bytes() ([]byte, error) {
+	buff := bytes.NewBuffer(nil)
+	err := ni.Serialize(buff)
+	if err != nil {
+		return nil, err
+	}
+	return buff.Bytes(), nil
 }
 
-func (sdb *StateDB) setName(registrant []byte, name string) {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) String() string {
+	return fmt.Sprintf("%064x %d", ni.registrant, ni.expiresAt)
+}
+
+func (ni *nameInfo) Empty() bool {
+	return len(ni.registrant) == 0 && ni.expiresAt == 0
+}
+
+func getRegistrantId(registrant []byte) string {
+	return hex.EncodeToString(registrant)
+}
+
+func getNameId(name string, registrant []byte, expiresAt uint32) ([]byte, error) {
+	nameId := bytes.NewBuffer(nil)
+	nameId.Write([]byte(name))
+
+	ni := newNameInfo(registrant, expiresAt)
+	niBytes, err := ni.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	registrantHash := sha256.Sum256(niBytes)
+	nameId.Write(registrantHash[:hashPrefixLength])
+	return nameId.Bytes(), nil
+}
+
+func (sdb *StateDB) getRegName(name string) (string, error) {
+	iter := trie.NewIterator(sdb.trie.NodeIterator(append(NamePrefix, name...)))
+	iter.Next()
+	data := iter.Value
+	if len(data) > 0 {
+		buff := bytes.NewBuffer(data)
+		ni := new(nameInfo)
+		if err := ni.Deserialize(buff); err != nil {
+			return "", err
+		}
+		return ni.String(), nil
+	}
 
-	sdb.names.Store(registrantId, name)
-	sdb.nameRegistrants.Store(nameId, registrant)
+	return "", errors.New("no name registration info found")
 }
 
-func (sdb *StateDB) deleteName(registrantId string) error {
-	name, err := sdb.trie.TryGet(append(NameRegistrantPrefix, registrantId...))
+func (sdb *StateDB) getNameInfo(nameId []byte) (*nameInfo, error) {
+	if v, ok := sdb.names.Load(string(nameId)); ok {
+		if ni, ok := v.(*nameInfo); ok {
+			return ni, nil
+		}
+	}
+
+	ni := new(nameInfo)
+	enc, err := sdb.trie.TryGet(append(NamePrefix, nameId...))
+	if err != nil {
+		return nil, err
+	}
+	if len(enc) > 0 {
+		ni := new(nameInfo)
+		buff := bytes.NewBuffer(enc)
+		if err := ni.Deserialize(buff); err != nil {
+			return nil, err
+		}
+		sdb.names.Store(string(nameId), ni)
+	}
+
+	return ni, nil
+}
+
+func (sdb *StateDB) nameRegistration(name string, registrant []byte, expiresAt uint32) error {
+	nameId, err := getNameId(name, registrant, expiresAt)
 	if err != nil {
 		return err
 	}
 
-	nameId := getNameId(string(name))
-	err = sdb.trie.TryDelete(append(NameRegistrantPrefix, registrantId...))
+	ni, err := sdb.getNameInfo(nameId)
 	if err != nil {
 		return err
 	}
 
-	err = sdb.trie.TryDelete(append(NamePrefix, nameId...))
+	if ni.Empty() {
+		ni.registrant = registrant
+		ni.expiresAt = expiresAt
+	} else {
+		return errors.New("name got registered")
```

Comment:
Suggest to rename to `name is already registered`

Reply:
We should probably allow the same registrant to extend already registered name by sending another `RegisterName` transaction.